File: src/tasks/cli/command/Command.java
package tasks.cli.command;

import static java.util.Objects.requireNonNull;

import omnia.data.structure.Collection;
import omnia.data.structure.List;
import omnia.data.structure.immutable.ImmutableList;
import omnia.data.structure.immutable.ImmutableSet;

public final class Command {
  private final String canonicalName;
  private final Collection<String> aliases;
  private final String description;
  private final Collection<Parameter> parameters;
  private final Collection<Option> options;

  private Command(
      String canonicalName,
      Collection<String> aliases,
      String description,
      Collection<Parameter> parameters,
      Collection<Option> options) {
    requireNonNull(canonicalName);
    requireNonNull(aliases);
    requireNonNull(description);
    requireNonNull(parameters);
    requireNonNull(options);

    if (aliases.contains(canonicalName)) {
      throw new IllegalArgumentException("aliases cannot contain the canonical name");
    }
    if (ImmutableSet.copyOf(aliases).count() < aliases.count()) {
      throw new IllegalArgumentException("aliases cannot contain duplicates: " + aliases);
    }

    this.canonicalName = canonicalName;
    this.aliases = ImmutableList.copyOf(aliases);
    this.description = description;
    this.parameters = parameters;
    this.options = ImmutableList.copyOf(options);
  }

  public String canonicalName() {
    return canonicalName;
  }

  public Collection<String> aliases() {
    return aliases;
  }

  public String description() {
    return description;
  }

  public Collection<Parameter> parameters() {
    return parameters;
  }

  public Collection<Option> options() {
    return options;
  }

  public List<String> canonicalNameAndAliases() {
    return ImmutableList.<String>builder().add(canonicalName).addAll(aliases).build();
  }

  public interface Builder0 {
    Builder1 canonicalName(String canonicalName);
  }

  public interface Builder1 {
    Builder2 aliases(String... aliases);
  }

  public interface Builder2 {
    Builder3 parameters(Collection<Parameter> parameters);
  }

  public interface Builder3 {
    Builder4 options(Collection<Option> options);
  }

  public interface Builder4 {
    Command helpDocumentation(String description);
  }

  public static Builder0 builder() {
    return canonicalName ->
        (Builder1) aliases ->
            (Builder2) parameters ->
                (Builder3) arguments ->
                    (Builder4) description ->
                        new Command(
                            canonicalName,
                            ImmutableList.copyOf(aliases),
                            description,
                            parameters,
                            arguments);
  }
}

File: cli/src/me/strangepan/tasks/cli/command/Command.kt
package me.strangepan.tasks.cli.command

import omnia.data.structure.Collection
import omnia.data.structure.List
import omnia.data.structure.immutable.ImmutableList
import omnia.data.structure.immutable.ImmutableSet

class Command private constructor(
    private val canonicalName: String,
    private val aliases: Collection<String>,
    private val description: String,
    private val parameters: Collection<Parameter>,
    private val options: Collection<Option>) {

  fun canonicalName(): String {
    return canonicalName
  }

  fun aliases(): Collection<String> {
    return aliases
  }

  fun description(): String {
    return description
  }

  fun parameters(): Collection<Parameter> {
    return parameters
  }

  fun options(): Collection<Option> {
    return options
  }

  fun canonicalNameAndAliases(): List<String> {
    return ImmutableList.builder<String>().add(canonicalName).addAll(aliases).build()
  }

  interface Builder0 {
    fun canonicalName(canonicalName: String): Builder1
  }

  interface Builder1 {
    fun aliases(vararg aliases: String): Builder2
  }

  interface Builder2 {
    fun parameters(parameters: Collection<Parameter>): Builder3
  }

  interface Builder3 {
    fun options(options: Collection<Option>): Builder4
  }

  interface Builder4 {
    fun helpDocumentation(description: String): Command
  }

  companion object {
    @JvmStatic
    fun builder(): Builder0 {
      return object : Builder0 {
        override fun canonicalName(canonicalName: String): Builder1 {
          return object : Builder1 {
            override fun aliases(vararg aliases: String): Builder2 {
              return object : Builder2 {
                override fun parameters(parameters: Collection<Parameter>): Builder3 {
                  return object : Builder3 {
                    override fun options(options: Collection<Option>): Builder4 {
                      return object : Builder4 {
                        override fun helpDocumentation(description: String): Command {
                          return Command(
                              canonicalName,
                              ImmutableList.copyOf(aliases),
                              description,
                              ImmutableList.copyOf(parameters),
                              ImmutableList.copyOf(options))
                        }
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
  }

  init {
    require(!aliases.contains(canonicalName)) { "aliases cannot contain the canonical name" }
    require(ImmutableSet.copyOf(aliases).count() >= aliases.count()) { "aliases cannot contain duplicates: $aliases" }
  }
}